Added `deftype` annotation for definitions database.

callendorph opened the pull request:
This works for both high and lo stanza but it only extracts the type arguments and the parent types. It doesn't try to extract any of the children content yet.

CuppoJava commented on compiler/stz-el.stanza:
```diff
@@ -130,6 +134,184 @@ defn ensure-unique-identifiers! (epackage:EPackage) :
+          EExternFn|EExtern|EDefType|EDefObject|EDefTypeObject) :
+    n(e)
+
+;; definitions that need to be in runtime and aren't traceable from roots
```
Please follow the code comment conventions. Comments begin with a single semicolon, are capitalized and end with periods.

jackbackrack replied:
done.

CuppoJava commented on compiler/stz-el.stanza:
```diff
@@ -130,6 +134,184 @@ defn ensure-unique-identifiers! (epackage:EPackage) :
+  do(walk-top, predefines(epackage))
+  while not empty?(q) : walk-top(pop(q))
+
+  println("FOUND %_ REACHABLE OUT OF %_ DEFINED %_%%" % [length(reachables), length(defs), 100.0 * to-double(length(reachables)) / to-double(length(defs))]) when prune-trace?
```
Please strip out development prints before merging.

jackbackrack replied:
done.

CuppoJava commented on compiler/stz-el.stanza:
```diff
@@ -130,6 +134,184 @@ defn ensure-unique-identifiers! (epackage:EPackage) :
+  val new-exps = remove-unused-var-stores $ for e in filter(keep?, exps(epackage)) seq : prune-methods(e, reachables)
+  val res = EPackage(new-packageio, to-tuple $ new-exps)
+
+  if prune-trace? :
```
Please strip out development prints before merging.

jackbackrack replied:
done.

CuppoJava commented on compiler/stz-stitcher.stanza:
```diff
@@ -362,12 +362,15 @@ public defn Stitcher (packages:Collection<VMPackage>, bindings:Bindings|False, s
-        val types* = map(resolve{pkgids, _}, types(m))
-        val branch* = Branch(to-list(types*), M(lbl))
-        add(method-table, multi*, branch*)
+        match(global-id(pkgids,multi(m))) :
```
Why does the tree-shaker affect the stripper?

jackbackrack replied:
because you can have methods without multis when stripping.  the stitcher assumed that methods always needed multis but if there are no calls to multis only to the methods then there is no need for a multi and in fact the multi doesn't exist.

CuppoJava replied:
Okay, thanks for the explanation. In this case, let's move the logic out of the stitcher, and promote the methods to simple functions in the final EL transform.

jackbackrack replied:
done.

callendorph committed:
Added `deftype` annotation for definitions database.

This works for both high and lostanza but it only extracts the
type arguments and the parent types. It doesn't try to extract
any of the children content yet.

CuppoJava commented:
Thanks! Looks great.
